sirepo/template/srw.py

```diff
@@ -335,6 +335,40 @@ def clean_run_dir(run_dir):
         zip_dir.remove()
 
 
+def create_archive(sim, qcall):
+    from sirepo import exporter
+
+    def _create_zip(sim, out_dir, qcall):
+        path = out_dir.join(sim.id + ".zip")
+        data = simulation_db.open_json_file(sim.type, sid=sim.id, qcall=qcall)
+        simulation_db.update_rsmanifest(data)
+        data.pkdel("report")
+        files = sirepo.sim_data.get_class(data).lib_files_for_export(data, qcall=qcall)
+        l = sim.filename.replace(".zip", "")
+        data.fDir = l
+        for f in exporter.python(data, sim, qcall):
+            files.append(f)
+        with sirepo.util.write_zip(str(path)) as z:
+            for f in files:
+                if f.basename not in ("run.py", "sirepo-data.json"):
+                    z.write(str(f), l + "/" + f.basename)
+                else:
+                    z.write(str(f), f.basename)
+            z.writestr(
+                simulation_db.SIMULATION_DATA_FILE,
+                pykern.pkjson.dump_pretty(data, pretty=True),
+            )
+        return path, data
+
+    with simulation_db.tmp_dir(qcall=qcall) as d:
+        f, c = _create_zip(sim, out_dir=d, qcall=qcall)
+        return qcall.reply_attachment(
+            f,
+            filename=sim.filename,
+            content_type="application/zip",
+        )
+
+
```

Comment:
> Rather than copy the export creation logic, what if the template was allowed to take the zipfile created by create_archive() and modify it? Instead, the call to the template would be right before the value was returned, and the name of the zipfile would be passed as an argument:
> 
> ```python
> if hasattr(sim.template, "create_archive"):
>     f = sim.template.create_archive(f)
> return qcall.reply_attachment(
>     f,
>     ...)
> ```
> 
> The logic of template.srw.create_archive would be to take the zipfile and move the members around - here is a hard-coded example which demonstrates renaming members and returning a new zip file
> 
> ```python
> import zipfile
> 
> with zipfile.ZipFile('in.zip', 'r') as zi:
>     with zipfile.ZipFile('out.zip', 'w') as zo:
>         for info in zi.infolist():
>             zo.writestr(info, zi.read(info))
>             if info.filename not in ('run.py', 'sirepo-data.json'):
>                 info.filename = f'NSLS-II_CHX_beamline_tabulated/{info.filename}'
>                 print('renamed: {}'.format(info.filename))
> return 'out.zip'
> ```
> 
> The code for setting data.fDir would be moved to template.srw.python_srouce_for_model, if model == None.

I'll try it out 👍

Reply:
It turns out that zip files keep 2 copies of each filename - one local and one central and you can't update both using python. Instead, the hook could be moved within exporter._create_zip():

```python
    #...
    with sirepo.util.write_zip(str(path)) as z:
        for f in files:
            if hasattr(sim.template, "export_filename"):
                n = sim.template.export_filename(sim.filename, f.basename)
            else:
                n = f.basename
            z.write(str(f), n)
    #...
````

and within template.srw, add export_filename():

```python
def export_filename(sim_filename, filename):
    if filename not in ("run.py", "sirepo-data.json"):
        sim_filename = sim_filename.replace(".zip", "")
        return f"{sim_filename}/{filename}"
    return filename
```
That way none of the exporter code needs to be copied.